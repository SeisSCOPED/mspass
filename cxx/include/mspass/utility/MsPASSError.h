#ifndef _MSPASS_ERROR_H_
#define _MSPASS_ERROR_H_
#include <iostream>
#include <exception>
namespace mspass {
using namespace std;
/*! \brief Severity code for error messages.

C++11 added the construct of a enum class that we use here.  It makes
the enum more strongly typed to avoid collisions for these common 
words that would be trouble otherwise.   This class is used in
all MsPASSError objects. */
enum class ErrorSeverity
{
	Fatal,
	Invalid,
	Suspect,
	Complaint,
	Debug,
	Informational
};
/*! \brief Base class for error object thrown by MsPASS library routines.

 This is the generic error object thrown by the MsPASS library.
 it is similar in concept to basic error objects described in various
 books by Stroustrup.  The base object contains only a simple
 generic message and a virtual log_error method common to all
 MsPASS error objects that are it's descendents.
\author Gary L. Pavlis
**/
class MsPASSError : public std::exception
{
public:
/*!
 Holds error message that can be printed with log_error method.
**/
	string message;
	ErrorSeverity badness;
/*!
 Default constructor built inline.
**/
	MsPASSError(){
		message="MsPASS library error\n";
		badness=ErrorSeverity::Fatal;
	};
/*! \brief Construct from a std::string with badness defined by keywords in a string.

Sometimes it is easier and more readable to use a string literal to define
the error class.  This uses that approach.

\param mess is the error message posted.
\param howbad is a string to translate to one of the allowed enum values.  The
  allowed values are the same as the enum defined in this file:
	FATAL,Invalid,Suspect,Complaint,Debug,Informational.
**/
	MsPASSError(const string mess,const char *howbad){
		message=mess;
		string s(howbad);
		if(s=="Fatal"){
		  badness=ErrorSeverity::Fatal;
		}
		else if (s=="Invalid"){
			badness=ErrorSeverity::Invalid;
		}
		else if (s=="Suspect"){
			badness=ErrorSeverity::Suspect;
		}
		else if (s=="Complaint"){
			badness=ErrorSeverity::Complaint;
		}
		else if (s=="Debug"){
			badness=ErrorSeverity::Debug;
		}
		else if (s=="Informational"){
			badness=ErrorSeverity::Informational;
		}
		else
		{
			/* Default to Fatal in this case as this should only happen in
			a debug situation. */
			badness=ErrorSeverity::Fatal;
		}
	};
/*! Construct from a string with enum defining severity.

This should be the normal form of this error object to throw.   Default of
the enum allows simpler usage for most errors.

\param mess - is the error message to be posted.
\param s is the severity enum (default Invalid).
*/
	MsPASSError(const string mess,const ErrorSeverity s=ErrorSeverity::Invalid)
        {
            message=mess;
            badness=s;
        };
/*! Construct from a char * and severity enum.
**/
	MsPASSError(const char *mess,const ErrorSeverity s){
		message=string(mess);
		badness=s;
	};
/*!
 Sends error message thrown by MsPASS library functions to standard error.
**/
void log_error(){
  cerr << message << endl;
};
/*! Overloaded method for sending error message to other than stderr. */
void log_error(ostream& ofs)
{
	ofs << message <<endl;
}
/*! This overrides the method in std::exception to load our string.
  This allows handlers to use the what method and get the error string
  from seisp.  Idea copied from:
  http://www.cplusplus.com/doc/tutorial/exceptions/
  */
  const char * what() const throw()
  {
    return message.c_str();
  };
	/*! Return error severity as the enum value. */
	ErrorSeverity severity(){return badness;};
};
}  // End mspass namespace declaration
#endif
