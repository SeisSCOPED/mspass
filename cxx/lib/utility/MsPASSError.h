#ifndef _MSPASS_ERROR_H_
#define _MSPASS_ERROR_H_
#include <iostream>
#include <exception>
namespace mspass {
using namespace std;
/*! \brief Base class for error object thrown by MsPASS library routines.

 This is the generic error object thrown by the MsPASS library.
 it is similar in concept to basic error objects described in various
 books by Stroustrup.  The base object contains only a simple
 generic message and a virtual log_error method common to all
 MsPASS error objects that are it's descendents.
\author Gary L. Pavlis
**/
class MsPASSError : public std::exception
{
public:
/*!
 Holds error message that can be printed with log_error method.
**/
	string message;
/*!
 Default constructor built inline.
**/
	MsPASSError(){message="MsPASS library error\n";};
/*! Construct from a std::string
**/
	MsPASSError(const string mess){message=mess;};
/*! Construct from a char *
**/
	MsPASSError(const char *mess){message=string(mess);};
/*!
 Sends error message thrown by MsPASS library functions to standard error.
**/
void log_error(){
  cerr << message << endl;
};
/*! Overloaded method for sending error message to other than stderr. */
void log_error(ostream& ofs)
{
	ofs << message <<endl;
}
/*! This overrides the method in std::exception to load our string.
  This allows handlers to use the what method and get the error string
  from seisp.  Idea copied from:
  http://www.cplusplus.com/doc/tutorial/exceptions/
  */
  const char * what() const throw()
  {
    return message.c_str();
  }
};
}  // End mspass namespace declaration
#endif
