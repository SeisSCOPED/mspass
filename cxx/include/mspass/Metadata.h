#ifndef _METADATA_H_
#define _METADATA_H_
#include <typeinfo>
#include <map>
#include <set>
#include <iostream>
#include <fstream>
#include <sstream>
#include <boost/any.hpp>
#include "MsPASSError.h"
#include "BasicMetadata.h"

namespace mspass
{
/*! \brief Error thrown when get operators fail.   
 *
 * This is a convenience class used to construct a more informative 
 * set of errors when get operations fail.  */
class MetadataGetError : public MsPASSError
{
public:
  stringstream ss;
  /*! Constructor called when a key is not found in the Metadata.
   * \param Texpected is the type name (return of typeid name method) trying to extract. */
  MetadataGetError(string key,string Texpected)
  {
    
    string pretty_name(boost::core::demangle(Texpected.c_str()));
    ss<<"Error trying to extract Metadata with key="<<key<<endl
      << "No value associated with this key is set in Metadata object"<<endl
      << "Expected an entry of type="<<pretty_name<<endl;
    message=ss.str();
  };
  /*! \brief Constructor called when type requested does not match contents. 
 
    This implementation uses a pickle style map container where the 
    contents of the map can by any type.  We use boost::any to provide
    sanity checks on types.   This is creates the error message thrown
    when the type of the return does not match the type requested. */
  MetadataGetError(const char *boostmessage,string key,
          string Texpected,string Tactual)
  {
    ss << "Error in Metadata get method.   Type mismatch in request"<<endl
      << "boost::any bad_any_cast wrote this message:  "<< boostmessage<<endl;
    string name_e(boost::core::demangle(Texpected.c_str()));
    ss << "Trying to convert to data of type="<<name_e<<endl;
    string name_a(boost::core::demangle(Tactual.c_str()));
    ss << "Actual entry has type="<<name_a<<endl;
    message=ss.str();
  };
};

class Metadata : public BasicMetadata
{
public:
  /*! Default constructor.  Does nothing.
  **/
  Metadata(){};
  /*! Construct from a file.

  This simple file based constructor assumes the file contains only a set
  lines with this format:   key value type
  where type must be one of:   real, integer, bool, or string.   Note int is actually
  always promoted to a long.  The optional format
  variable is there to allow alternative formats in the future.

  \param ifs - ifstream from which to read data
  \param format - optional format specification.   Currently only default of "text"
     is accepted.
  \exception MsPASSError can be thrown for a variety of conditions.
  */
  Metadata(ifstream& ifs, const string form=string("pf"));
  /*!
  Standard copy constructor.

  \param mdold - parent object to be copied
  **/
  Metadata(const Metadata& mdold);
  /*! Standard assignment operator.
    \param mdold - parent object to copy
  */
  Metadata& operator=(const Metadata& mdold);
  /*! Append additional metadata with replacement.

A plus operator implies addition, but this overloading does something very
different.  A simple way to describe the effect is that on completion the
left hand side Metadata object will contain a duplicate of the right hand
side plus any attributes in the rhs that were not present on the lhs.
Another way to clarify this is to describe the algorithm.   We take each
attribute on the right and search for it in the lhs.  If it is not in the lhs
it will be added.  If it is there already, the rhs value will replace the old
value on the lhs.   This is most useful when an algorithm creates a new set of
attributes that we want to use in downstream processing but retain all the
other attributes.

\param rhs is the new metadata to be insert/replace on the lhs.
*/
  Metadata& operator+=(const Metadata& rhs) noexcept;
  const Metadata operator+(const Metadata& other) const;
  /* All the getters - all but the template are wrappers with the type
  fixed */
  /*!
  Get a real number from the Metadata object.

  \exception MetadataGetError if requested parameter is not found or there is a type mismatch.
  \param key keyword associated with requested metadata member.
  **/
  double get_double(const string key){
    try{
      double val;
      val=get<double>(key);
      return val;
    }catch(...){throw;};
  };
  /*!
  Get an integer from the Metadata object.

  \exception MetadataGetError if requested parameter is not found or there is a type mismatch.
  \param key keyword associated with requested metadata member.
  **/
  int get_int(const string key)
  {
    try{
      int val;
      val=get<int>(key);
      return val;
    }catch(...){throw;};
  };
  /*!
  Get a long integer from the Metadata object.

  \exception MetadataGetError if requested parameter is not found or there is a type mismatch.
  \param key keyword associated with requested metadata member.
  **/
  long get_long(const string key){
    try{
      long val;
      val=get<long>(key);
      return val;
    }catch(...){throw;};
  };
  /*!
  Get a string from the Metadata object.

  Note the string in this case can be quite large.  If the string
  was parsed from an Antelope Pf nested Tbl and Arrs can be extracted
  this way and parsed with pf routines.

  \exception MetadataGetError if requested parameter is not found or there is a type mismatch.
  \param key keyword associated with requested metadata member.
  **/
  string get_string(const string key){
    try{
      string val;
      val=get<string>(key);
      return val;
    }catch(...){throw;};
  };
  /*!
  Get a  boolean parameter from the Metadata object.

  This method never throws an exception assuming that if the
  requested parameter is not found it is false.

  \param key keyword associated with requested metadata member.
  **/
  bool get_bool(const string key){
    try{
      bool val;
      val=get<bool>(key);
      return val;
    }catch(...){throw;};
  };
  /*! Generic get interface.

  This is a generic interface most useful for template procedures
  that need to get a Metadata component.   Since this object only
  can contain simple types the type requested must be simple.
  Currently supports only int, long, short, double, float, and string.
  C char* is intentionally not supported.
  Calls to anything but the supported types will
  throw an exception.

  \param key is the name tag of desired component.

  \exception - will throw a MetadataGetError (child of MsPASSError) for
       type mismatch or in an overflow or underflow condition.
  */
  template <typename T> T get(const string key);
  /*! \brief Generic get interface for C char array.

          This is a generic interface most useful for template procedures
          that need to get a Metadata component.   Since this object only
          can contain simple types the type requested must be simple.
          Currently supports only int, long, short, double, float, and string.
          C char* is intentionally not supported. This is largely a wrapper
          on the string key version of this same generic function.

          \param key is the name tag of desired component.

          \exception - will throw a MetadataGetError (child of MsPASSError) for
             type mismatch or in an overflow or underflow condition.
             */
  template <typename T> T get(const char *key)
  {
    try{
      T val;
      val=get<T>(string(key));
      return val;
    }catch(...){throw;};
  }
  template <typename T> void put(const string key, T val) noexcept
  {
    boost::any aval=val;
    md[key]=aval;
    changed_or_set.insert(key);
  }
  template <typename T> void put (const char *key, T val) noexcept
  {
    /* could do this as put(string(key),val) but this is so trivial duplicating
    the code for the string method is more efficient than an added function call.*/
    boost::any aval=val;
    md[string(key)]=aval;
    changed_or_set.insert(string(key));
  }
  void put(const string key, const double val)
  {
      this->put<double>(key,val);
  };
  void put(const string key, const int val)
  {
      this->put<int>(key,val);
  };
  void put(const string key, const bool val)
  {
      this->put<bool>(key,val);
  };
  void put(const string key, const string val)
  {
      this->put<string>(key,val);
  };

  friend ostream& operator<<(ostream&, Metadata&);
protected:
  map<string,boost::any> md;
  /* The keys of any entry changed will be contained here.   */
  set<string> changed_or_set;
};

template <typename T> T Metadata::get(const string key)
{
  T result;
  map<string,boost::any>::const_iterator iptr;
  iptr=md.find(key);
  if(iptr==md.end())
  {
    throw MetadataGetError(key,typeid(T).name());
  }
  boost::any aval=iptr->second;
  try{
    result=boost::any_cast<T>(aval);
  }catch(boost::bad_any_cast& err)
  {
    const std::type_info &ti = aval.type();
    throw MetadataGetError(err.what(),key,typeid(T).name(),ti.name());
  };
  return result;
}
}  //End of namespace MsPASS
#endif
