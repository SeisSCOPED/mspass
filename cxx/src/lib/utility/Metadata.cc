#include <boost/core/demangle.hpp>
#include "mspass/utility/Metadata.h"
namespace mspass
{
Metadata::Metadata(ifstream& ifs, const string form)
{
  try{
    char linebuffer[256];
    while(ifs.getline(linebuffer,128))
    {
      string s1,s2,s3;
      boost::any a;
      stringstream ss(linebuffer);
      ss>>s1;
      ss>>s2;
      ss>>s3;
      if(s3=="real")
      {
        double dval;
        dval=atof(s2.c_str());
        a=dval;
        md[s1]=a;
        changed_or_set.insert(s1);
      }
      else if(s3=="integer")
      {
        long ival;
        ival=atol(s2.c_str());
        a=ival;
        md[s1]=a;
        changed_or_set.insert(s1);
      }
      else if(s3=="string")
      {
        string sval;
        a=sval;
        md[s1]=a;
        changed_or_set.insert(s1);
      }
      else if(s3=="bool")
      {
        bool bval;
        if( (s2=="TRUE")||(s2=="true")||(s2=="1") )
          bval=true;
        else
          bval=false;
        a=bval;
        md[s1]=a;
        changed_or_set.insert(s1);
      }
      else
      {
        stringstream sserr;
        sserr << "Metadata file constructor:  Illegal type specification for key="
          << s1<<" with a value field of "<<s2<<endl
          << "type specified as "<<s3<<" is illegal.  "
          << "Must be on of the following:  real, integer, bool, or string."<<endl;
        throw MsPASSError(sserr.str(),ErrorSeverity::Invalid);
      }
    }
  }catch(...){throw;};
}
Metadata::Metadata(const Metadata& parent)
  : md(parent.md),changed_or_set(parent.changed_or_set)
{
}

Metadata& Metadata::operator=(const Metadata& parent)
{
  if(this!=(&parent))
  {
    md=parent.md;
    changed_or_set=parent.changed_or_set;
  }
  return *this;
}

Metadata& Metadata::operator+=(const Metadata& rhs) noexcept
{
  if(this!=(&rhs))
  {
    /* We depend here upon the map container replacing values associated with
    existing keys.  We mark all entries changes anyway.  This is a vastly simpler
    algorithm than the old SEISPP::Metadata.  */
    map<string,boost::any>::const_iterator rhsptr;
    for(rhsptr=rhs.md.begin();rhsptr!=rhs.md.end();++rhsptr)
    {
      md[rhsptr->first]=rhsptr->second;
      changed_or_set.insert(rhsptr->first);
    }
  }
  return *this;
}
const Metadata Metadata::operator+(const Metadata& other) const
{
  Metadata result(*this);
  result += other;
  return result;
}
set<string> Metadata::keys() noexcept
{
  set<string> result;
  map<string,boost::any>::iterator mptr;
  for(mptr=md.begin();mptr!=md.end();++mptr)
  {
    string key(mptr->first);\
    result.insert(key);
  }
  return result;
}
/* Helper returns demangled name using boost demangle.  */
string demangled_name(boost::any a)
{
    try{
        const std::type_info &ti = a.type();
        const char *rawname=ti.name();
        string pretty_name(boost::core::demangle(rawname));
        return pretty_name;
    }catch(...){throw;};
}
ostream& operator<<(ostream& os, Metadata& m)
{
  try{
    map<string,boost::any>::iterator mdptr;
    for(mdptr=m.md.begin();mdptr!=m.md.end();++mdptr)
    {
        /* Only handle simple types for now.  Issue an error message to 
         * cerr for other types */
        int ival;
        long lval;
        double dval;
        float fval;
        string sval;
        bool bval;
        boost::any a=mdptr->second;
        /* A relic retained to help remember this construct*/
        //const std::type_info &ti = a.type();
        string pretty_name=demangled_name(a);
        /*WARNING:  potential future maintance issue.
         * Currently the demangling process is not standardized and the
         * boost code used here does not return a name that is at all 
         * pretty for string data.   This crude approach just tests for
         * the keyword basic_string embedded in the long name.  This works
         * for now, but could create problems if and when this anomaly
         * evolves away. */
        string sname("string");
        if(pretty_name.find("basic_string")==std::string::npos)
            sname=pretty_name;
        os<<mdptr->first<<" "<<sname<<" ";
        try{
            if(sname=="int")
            {
                ival=boost::any_cast<int>(a);
                os<<ival<<endl;
            }
            else if(sname=="long")
            {
                lval=boost::any_cast<long int>(a);
                os<<lval<<endl;
            }
            else if(sname=="double")
            {
                dval=boost::any_cast<double>(a);
                os<<dval<<endl;
            }
            else if(sname=="float")
            {
                fval=boost::any_cast<float>(a);
                os<<fval<<endl;
            }
            else if(sname=="bool")
            {
                bval=boost::any_cast<bool>(a);
                os<<bval<<endl;
            }
            else if(sname=="string")
            {
                sval=boost::any_cast<string>(a);
                os<<sval<<endl;
            }
            else
            {
                os <<"NONPRINTABLE"<<endl;
            }
        }catch(boost::bad_any_cast &e)
        {
            os<<"BAD_ANY_CAST_ERROR"<<endl;
        }
    }
    return os;
  }catch(...){throw;};
}

} // End mspass Namespace block
