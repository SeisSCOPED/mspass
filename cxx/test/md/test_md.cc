#include "mspass/utility/ErrorLogger.h"
#include "mspass/utility/MsPASSError.h"
#include "mspass/utility/Metadata.h"
#include "mspass/utility/AntelopePf.h"
using namespace mspass;
int main(int argc, char **argv)
{
	char *pfname=strdup("test_md");
	try {
            cout << "Test program for Metadata object, AntelopePf, and ErrorLogger"
                <<endl
            << "First try creating an ErrorLogger object"<<endl;
            ErrorLogger elog;
            cout << "Try algorithms to set jobid and algorithm"<<endl;
            elog.set_job_id(10000);
            elog.set_algorithm(string("test_md"));
            cout << "Success - job_id set to "<<elog.get_job_id()<<endl
                << "Algorithm="<<elog.get_algorithm()<<endl;
		cout << "Trying to build Metadata objects" << endl;
		cout << "Trying default constructor" << endl;
		Metadata mdplain;
                cout << "Trying put methods for int, double, string, and bool"
                    <<endl;
                long lval; double dval; string sval;  bool bval;
                lval=10;
                mdplain.put<long>("long_val",lval);
                dval=2.5;
                mdplain.put<double>("double_val",dval);
                sval=string("test_string");
                mdplain.put<string>("string_val",sval);
                bval=true;
                mdplain.put<bool>("bool_val",bval);
                cout << "Succeeded - trying matching get methods"<<endl;
                cout << "long_val="<<mdplain.get<long>("long_val")<<endl;
                cout << "double_val="<<mdplain.get<double>("double_val")<<endl;
                cout << "string_val="<<mdplain.get<string>("string_val")<<endl;
                cout << "bool_val="<<mdplain.get<bool>("bool_val")<<endl;
                cout << "Same thing using operator >> "<<endl;
		cout << mdplain;
		cout << "Trying simple file read constructor"<<endl
                    << "Reading from simple.txt"<<endl;
                ifstream ifs("simple.txt");
                Metadata mds(ifs);
		cout << mds;
                cout << "Trying to read more complex pf file using AntelopePF object constructor"<<endl;
                AntelopePf pfsmd(pfname);
                cout << "Success - read the following:  "<<endl
                    << pfsmd<<endl;
                cout << "Trying assignment operator for Metadata with RTTI"<<endl;
                Metadata mdsum;
                mdsum=dynamic_cast<Metadata&>(pfsmd);
                cout << "Worked"<<endl<<"Contents of copy (simple attributes ony)"<<endl;
                cout << mdsum<<endl;
                cout << "Trying += operator.  Merging inline and pf objects"<<endl;
                cout << "Trying to add simple to Metadata derived from pf"<<endl;
                mdsum+=mds;
                cout << "Done - result:"<<endl;
                cout << mdsum<<endl;
                cout << "Reading and writing a couple of simple parameters"<<endl;
                cout << "simple_real_parameter="
                    <<pfsmd.get<double>("simple_real_parameter")<<endl
                  << "simple_int_parameter="
                    <<pfsmd.get<int>("simple_int_parameter")<<endl
                  << "simple_bool_parameter="<<pfsmd.get_bool("simple_bool_parameter")
                  <<endl;
                cout << "Trying get_tbl method"<<endl;
                list<string> tsttbl;
                tsttbl=pfsmd.get_tbl("mdlist");
                cout << "mdlist Tbl extracted from pf"<<endl;
                list<string>::iterator iptr;
                for(iptr=tsttbl.begin();iptr!=tsttbl.end();++iptr)
                {
                  cout << *iptr<<endl;
                }
                cout << "Trying get_branch method - extracting test_nested_tag"
                  <<endl;
                AntelopePf pfbr(pfsmd.get_branch("test_nested_tag"));
                cout << "Success"<<endl
                    <<"Contents"<<endl
                    << dynamic_cast<Metadata&>(pfbr)<<endl;
                cout << "test_double parameter in branch="<<pfbr.get_double("test_double")<<endl;
                cout << "Testing exceptions.  First a get failure:"<<endl;
                try{
                    double dbad=mdsum.get<double>("bad_key");
                    cout << "PROBLEM:  get did not throw error and returend"
                        << dbad<<endl;
                }catch(MetadataGetError& mdge)
                {
                    cout << "Properly handled.  Message posted follows:"<<endl
                        << mdge.what()<<endl;
                    if(mdge.severity() == ErrorSeverity::Invalid)
                        cout << "Correct severity of Invalid was posted"<<endl;
                    else 
                        cout << "Error - severity posted was not Invalid as expected"
                            <<endl;
                    cout << "Trying to write log with ErrorLogger"<<endl;
                    elog.log_error(mdge);
                    elog.log_verbose(string("log_error method succeeded"));
                }
                cout << "Trying intentional type mismatch."<<endl;
                try{
                    int ibad;
                    ibad=pfbr.get<int>("test_double");
                }catch(MetadataGetError& mdge)
                {
                    cout << "Properly handled trying to get test_double as int"
                        <<endl
                        <<"Error message posted"<<endl
                        << mdge.what()<<endl;
                    if(mdge.severity() == ErrorSeverity::Invalid)
                        cout << "Correct severity of Invalid was posted"<<endl;
                    else 
                        cout << "Error - severity posted was not Invalid as expected"
                            <<endl;
                    cout << "Posting that message to log"<<endl;
                    elog.log_error(mdge);
                }
                cout << "Posting a set of fake messages to log"<<endl;
                MsPASSError efatal(string("Fake fatal error"),"Fatal");
                elog.log_error(efatal);
                MsPASSError edebug(string("Fake debug error"),"Debug");
                elog.log_error(edebug);
                MsPASSError ec(string("Fake complain error"),"Complaint");
                elog.log_error(ec);
                list<LogData> ldata=elog.get_error_log();
                list<LogData>::iterator lptr;
                cout << "Dump of error log"<<endl;
                for(lptr=ldata.begin();lptr!=ldata.end();++lptr)
                {
                    cout << *lptr<<endl;
                }
	}
	catch (MsPASSError& sess)
	{
            cout << "Something threw an MsPASSError exception - message posted follows"<<endl;
	    cout << sess.what()<<endl;
            if(sess.severity() == ErrorSeverity::Invalid)
                cout << "Correct severity of Invalid was posted"<<endl;
            else 
                cout << "Error - severity posted was not Invalid as expected"
                    <<endl;
	}
        catch(std::exception& stex)
        {
            cout << "Something threw a std::exception that was not a MsPASSError"<<endl
                << "Error message:  "<<stex.what()<<endl;
	
        }
}
