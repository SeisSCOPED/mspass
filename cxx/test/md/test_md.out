Trying to build Metadata objects
Trying default constructor
Trying put methods for int, double, string, and bool
Succeeded - trying matching get methods
long_val=10
double_val=2.5
string_val=test_string
bool_val=1
Same thing using operator >> 
bool_val bool 1
double_val double 2.5
long_val long 10
string_val string test_string
Trying simple file read constructor
Reading from simple.txt
MYPAR double 2.5
cpar string 
i long 27
nsamp long -999
nx long 10
ny long 4
testbool bool 1
teststring string 
x double 4.5
y double 200
Trying to read more complex pf file using AntelopePF object constructor
Success - read the following:  
simple_bool_parameter bool 1
simple_int_parameter int 3
simple_real_parameter double 2
simple_string_parameter string test_string

Trying assignment operator for Metadata with RTTI
Worked
Contents of copy (simple attributes ony)
simple_bool_parameter bool 1
simple_int_parameter int 3
simple_real_parameter double 2
simple_string_parameter string test_string

Trying += operator.  Merging inline and pf objects
Trying to add simple to Metadata derived from pf
Done - result:
MYPAR double 2.5
cpar string 
i long 27
nsamp long -999
nx long 10
ny long 4
simple_bool_parameter bool 1
simple_int_parameter int 3
simple_real_parameter double 2
simple_string_parameter string test_string
testbool bool 1
teststring string 
x double 4.5
y double 200

Reading and writing a couple of simple parameters
simple_real_parameter=2
simple_int_parameter=3
simple_bool_parameter=1
Trying get_tbl method
mdlist Tbl extracted from pf
sta string
chan string
time real
endtime real
nsamp int
samprate real
Trying get_branch method - extracting test_nested_tag
Success
Contents
test_double double 2
test_int int 4
test_multi_word_string string This is a string with blanks.
test_string string foo

test_double parameter in branch=2
Testing exceptions.  First a get failure:
Properly handled.  Message posted follows:
Error trying to extract Metadata with key=bad_key
No value associated with this key is set in Metadata object
Expected an entry of type=double

Trying intentional type mismatch.
Properly handled trying to get test_double as int
Error message posted
Error in Metadata get method.   Type mismatch in request
boost::any bad_any_cast wrote this message:  boost::bad_any_cast: failed conversion using boost::any_cast
Trying to convert to data of type=int
Actual entry has type=double

